nats_test.go

```diff
@@ -276,6 +276,63 @@ var testServers = []string{
 	"nats://localhost:1228",
 }
 
+func TestMaxConnectionsReconnect(t *testing.T) {
+
+	// Start first server
+	s1Opts := natsserver.DefaultTestOptions
+	s1Opts.Port = -1
+	s1Opts.MaxConn = 2
+	s1Opts.Cluster = server.ClusterOpts{Name: "test", Host: "127.0.0.1", Port: -1}
+	s1 := RunServerWithOptions(&s1Opts)
+	defer s1.Shutdown()
+
+	// Start second server
+	s2Opts := natsserver.DefaultTestOptions
+	s2Opts.Port = -1
+	s2Opts.MaxConn = 2
+	s2Opts.Cluster = server.ClusterOpts{Name: "test", Host: "127.0.0.1", Port: -1}
+	s2Opts.Routes = server.RoutesFromStr(fmt.Sprintf("nats://127.0.0.1:%d", s1Opts.Cluster.Port))
+	s2 := RunServerWithOptions(&s2Opts)
+	defer s2.Shutdown()
+
+	opts := []Option{
+		MaxReconnects(2),
+		ReconnectWait(10 * time.Millisecond),
+		Timeout(200 * time.Millisecond),
+	}
+
+	// Create two connections (the current max) to first server
+	nc1, _ := Connect(s1.ClientURL(), opts...)
+	defer nc1.Close()
+	nc1.Flush()
+
+	nc2, _ := Connect(s1.ClientURL(), opts...)
+	defer nc2.Close()
+	nc2.Flush()
+
+	if s1.NumClients() != 2 {
+		t.Fatalf("Expected 2 client connections to first server. Got %d", s1.NumClients())
+	}
+
+	if s2.NumClients() > 0 {
+		t.Fatalf("Expected 0 client connections to second server. Got %d", s2.NumClients())
+	}
+
+	// Kick one of our two server connections off first server. One client should reconnect to second server
+	newS1Opts := s1Opts
+	newS1Opts.MaxConn = 1
+	err := s1.ReloadOptions(&newS1Opts)
+	if err != nil {
+		t.Fatalf("Unexpected error changing max_connections [%s]", err)
+	}
+
+	time.Sleep(200 * time.Millisecond)
```

Comment:
To avoid the sleep here, you could maybe use the reconnect / disconnect handlers in the options like this:

```go
	reconnectCh := make(chan struct{})
	errCh := make(chan error, 2)
	opts := []Option{
		MaxReconnects(2),
		ReconnectWait(10*time.Millisecond),
		Timeout(200*time.Millisecond),
		ErrorHandler(func(_ *Conn, _ *Subscription, err error) {
			fmt.Println("ERROR: ", err)
		}),
		DisconnectErrHandler(func(_ *Conn, err error) {
			if err != nil {
				errCh <- err
			}
		}),
		ReconnectHandler(func(_ *Conn){
			reconnectCh <- struct{}{}
		}),
	}
```

Then replace the sleep with a couple of channels that check for the disconnect error and reconnect event:

```go
	select {
	case err := <-errCh:
		if err != ErrMaxConnectionsExceeded {
			t.Fatalf("Unexpected error %v", err)
		}
	case <-time.After(2*time.Second):
		t.Fatal("Timed out waiting for disconnect event")
	}

	select {
	case <-reconnectCh:
	case <-time.After(2*time.Second):
		t.Fatal("Timed out waiting for reconnect event")
	}
```

Reply:
Very good recommendation: you will both check that the error was what you expected *and* remove any flapping condition due to 200ms being maybe not enough some time or too much..

Reply:
Good suggestion @wallyqs .  Those hard-coded sleeps do seem like ticking time bombs (or at least wasted time).

Reply:
Change made and added to PR.